### DISCUSSION TITLE
Cypress Studio Experimental Feedback

### DISCUSSION THREAD

**Discussion by panzarino** [cypress-io/cypress#14339] [General]
Hey everyone, thanks so much for testing out the early experimental version of Cypress Studio!

We really appreciate any and all feedback that you have to provide. Please use this thread as a place to provide feedback, request features, or just comment on Cypress Studio in general.

If you have a larger issue to report (and it's not already listed [here](https://github.com/cypress-io/cypress/issues?q=is%3Aissue+label%3A%22experiment%3A+studio%22)), please [file a dedicated issue](https://github.com/cypress-io/cypress/issues/new?assignees=&labels=experiment%3A%20studio&template=1-bug-report.md&title=).

Some features that are in development:
- Assertions

Some features that we're considering releasing in the future:
- Run a test before saving

Features that have been added since initial release:
- Create a new test in an empty spec file
- Create a new spec file from Cypress
- Copy generated commands to clipboard
- Various improvements and enhancements

We're super excited about releasing this feature and can't wait to see the tests that you create with it!

[Studio Issue Tracker](https://github.com/cypress-io/cypress/issues?q=is%3Aissue+label%3A%22experiment%3A+studio%22)

[Studio Documentation](https://docs.cypress.io/guides/core-concepts/cypress-studio.html)

**Comment by estefafdez** [comment]
Is there any way to test this? I would love to give it a try! thanks!

**Reply by panzarino** [reply]
@estefafdez We'll be releasing it behind an experimental flag very soon - this thread was made in anticipation so we could set up links. I'll ping here once it actually goes out

**Reply by panzarino** [reply]
@estefafdez We've just released this feature behind an experimental flag in 6.3.0 - you can check out the docs [here](https://docs.cypress.io/guides/core-concepts/cypress-studio.html).

**Comment by dverbiest** [comment]
Some first reactions/feedback:
- Awesome to have such a feature in Cypress!! Great design, integration in the current Desktop as well 🥇 
- Can you please include a toggle to find and generate cy.contains('human readable button text') as well?
- I got some cases where the found selector does not work, e.g. because it selects on a .active class which is only there while clicking on the button, and not beforehand.
- Add an option to type/edit the found selectors yourself, maybe by using the input field of the Selector Playground? (already supports get and contains 👍 
- Add the pause and step by step options when Studio is running?
- te be continued... :)

**Reply by panzarino** [reply]
Thanks for the feedback! The selector problem (`.active`) is definitely something that needs to be fixed and I'll most likely be able to get that updated before the next release - I've created an issue here #14658.

Adding `.contains` as an autogenerated selector could be a little more complicated - we don't currently support that in the selector playground yet (you can only test what elements `.contains` will find, not the reverse). If requested enough we can invest some time into this and add it to both the selector playground and Studio.

I'm not quite sure what you mean by step by step option - are you talking about the snapshot that displays when you hover over a command in the log?

**Reply by dverbiest** [reply]
Thanks for the quick reply.

What I meant was: In a test run you can pause the test with cy.pause()
( by the way: why isn't there a pause button in the runner? :thinking: )
... and then step through the next commands, step by step using:

![image](https://user-images.githubusercontent.com/30799812/105224920-1efeca80-5b5e-11eb-952d-159a407a264c.png)


This option is not there when you rerun the test while working inside the Studio,
although I think it would be especially useful there.

**Reply by panzarino** [reply]
Got it - I'll definitely look into adding this for the next release or perhaps a bit later. We're planning on adding the ability for you to do a "test run" of your test before saving it and I think this would be super useful there as well.

**Comment by jimpriest** [comment]
I like the basic implementation!   Limited commands so can't do much but the basics are there.  
Need ability to edit entries. 
Would be great if you were given a few choices on which selector to use or the ability to enter one yourself. 

I was able to add commands but trying to save just gave me a vague "Cypress was unable to save these commands to your spec file."  

But looking forward to seeing how this feature grows!

**Reply by ChaseBig** [reply]
Trying the Studio beta feature on my Mac running the latest OSX. 

I'm also getting this error no matter what I try -- `Cypress was unable to save these commands to your spec file. Cypress Studio is still in beta and the team is working hard to resolve issues like this. To help us fix this issue more quickly, you can provide us with more information by clicking 'Learn more' below.`

Are there any configurations I can change or accessibility settings I need to change in order for this to work?

**Reply by panzarino** [reply]
So this error is most likely happening because sourcemaps aren't being generated properly and Studio is having a hard time finding the correct spot in your code to save the commands - there's a couple things that we can do to make this more resilient even without sourcemaps that we can work on for the next release.

Are you using Typescript? We had a couple issues when testing this internally but thought they were mostly all resolved. Regardless of whether or not you're using ts, if you're able to share your tsconfig and webpack config that would really help us out in reproducing the issue.

**Reply by ChaseBig** [reply]
Figured out what was causing my issue. The folder containing spec test files cannot have whitespace. After updating the name of the folder holding the test from `Local Dev` to `localDev`, the feature ended up working. 

This was the console message that lead me to believe the folder whitespace was causing the issue: 
`Autofocus processing was blocked because a document's URL has a fragment '#/tests/integration/dev_tests/Local%20Dev/fake_dev_test.spec.js'.`

**Reply by panzarino** [reply]
Thanks - that's super helpful for reproducing. I've filed an issue here #14659.

**Comment by davincho** [comment]
Hey, thanks so much for create Cypress - writing tests with Cypress are a game changer.

It would be cool to be able to hook into the code when the "best" selector is being determined.
To give you a little bit more context, I am using https://github.com/testing-library/cypress-testing-library and instead of querying a button with `cy.get` (how it is currently proposed by Cypress Studio), I would like to be able to return `cy.getByRole` leveraging `getSuggestedQuery` from https://github.com/testing-library/dom-testing-library.

**Reply by panzarino** [reply]
@davincho Thanks for the feedback! You can currently customize the way that selectors are generated by configuring the [Selector Playground](https://docs.cypress.io/api/cypress-api/selector-playground-api.html#Arguments) (this will be documented under Cypress Studio soon, too.). You can actually specify a custom function for generating a selector if needed, but unfortunately it will still have to work with `.get()` as Studio is only going to generate official Cypress commands. We are still planning on adding support for additional commands in subsequent releases.

In the future it's possible that we may add a plugin system that makes it easier to "hook in" to different actions within Studio (and someone could add support for the 3rd party commands you mentioned). However, this is still an early experimental release so any support for a plugin system is likely far away.

**Reply by hansOSA** [reply]
Thanks for linking the Selector Playground! We've been using data-cyid as a selector in our own app, I didn't realize we just had to change to data-cy to make the Cypress Studio super helpful!

**Comment by hman61** [comment]
It would be great if during the experimental phase (or permanently) could output the recorded steps in "raw" format to the log or to an expandable textarea in the left pane, either in csv format or line by line with the number being the test step delimiter which you already are logging in the left pane.

**Sample:**
get,#new-todo,click
get,#new-todo,type,coolio{enter}

 --- OR ---
2
get
#new-todo
click
3
get
#new-todo
type
coolio{enter}
4
get
.toggle
check

**Reply by hman61** [reply]
This would be very helpful until you have fixed the saving issue, because I'm currently using a framework that converts csv to cypress code, so I could already leverage the recording results if they were outputted in csv format

**Reply by panzarino** [reply]
@hman61 Thanks for the feedback! Is there a chance that the saving issue you're having has to do with spaces or special characters in the file name (#14659, #14679)? If so, that's been fixed and should be getting released this Monday. If that's not the cause of your issues, could you please provide some additional information in order to help us reproduce the save issue that you're having?

We've explored potentially outputting the code with a `Copy to Clipboard` button directly in the left pane, but we still haven't decided whether or not we'll be adding that during this experimental phase. I know you mention a csv format, but since most of our users don't use that tool we would most likely output the commands as a code block (in the format `cy.get('#new-todo').click()`. Would you still find that useful? If saving to file was working correctly for you, would you ever want to still copy the commands manually or would you just save it to file?

**Reply by hman61** [reply]
My spec name had spaces in it (doh), I tried another one without spaces and it saved fine. Thanks for getting back to me so quickly.

I still think the "Copy to Clip Board" would be very valuable because then you have the flexibility  to modify it by adding some existing @aliases and maybe paste the command steps to an existing function or a before each block instead of just using the command steps in the current test spec

**Reply by panzarino** [reply]
@hman61 Happy to help! We definitely have copy to clipboard on our radar along with editing `beforeEach` hooks and stuff along those lines.

**Reply by panzarino** [reply]
We've just released copy to clipboard in v7.7.0 - check it out and let us know what you think!

**Comment by ibsus** [comment]
It would be very helpful if there was a method to capture requests and turn them into fixtures.

**Reply by powerdong** [reply]
@panzarino  I want to know how to implement it?

**Comment by hman61** [comment]
I'm looking for an elegant way to be-able to read csv files that describes "dynamic tests" from fixtures prior to tests are being loaded. I know I can do it by adding a csv loader to the webpack config, or I can describe what need in json format and just use require('my.json') but the foot print is so much smaller when you describe the test in csv vs. json and much much faster to author in csv format.

**Comment by woutje74** [comment]
When attempting to provide feedback via the 'Give Feedback' button in the Testrunner, Cypress shuts down if you're running the test in the Electron browser. 
TypeError: Cannot read property 'show' of undefined
    at Object._launch (/Users/User/Library/Caches/Cypress/6.4.0/Cypress.app/Contents/Resources/app/packages/server/lib/browsers/electron.js:193:17)
    at Object._launchChild (/Users/User/Library/Caches/Cypress/6.4.0/Cypress.app/Contents/Resources/app/packages/server/lib/browsers/electron.js:189:17)
    at BrowserWindow.onNewWindow (/Users/User/Library/Caches/Cypress/6.4.0/Cypress.app/Contents/Resources/app/packages/server/lib/browsers/electron.js:137:22)
    at Object.<anonymous> (/Users/User/Library/Caches/Cypress/6.4.0/Cypress.app/Contents/Resources/app/packages/server/lib/gui/windows.js:192:36)
    at Object.emit (events.js:315:20)
    at Object.internalWindowOpen (electron/js2c/browser_init.js:181:2714)
    at Object.<anonymous> (electron/js2c/browser_init.js:157:11014)
    at Object.emit (events.js:315:20)

TypeError: Cannot read property 'show' of undefined
    at Object._launch (/Users/User/Library/Caches/Cypress/6.4.0/Cypress.app/Contents/Resources/app/packages/server/lib/browsers/electron.js:193:17)
    at Object._launchChild (/Users/User/Library/Caches/Cypress/6.4.0/Cypress.app/Contents/Resources/app/packages/server/lib/browsers/electron.js:189:17)
    at BrowserWindow.onNewWindow (/Users/User/Library/Caches/Cypress/6.4.0/Cypress.app/Contents/Resources/app/packages/server/lib/browsers/electron.js:137:22)
    at Object.<anonymous> (/Users/User/Library/Caches/Cypress/6.4.0/Cypress.app/Contents/Resources/app/packages/server/lib/gui/windows.js:192:36)
    at Object.emit (events.js:315:20)
    at Object.internalWindowOpen (electron/js2c/browser_init.js:181:2714)
    at Object.<anonymous> (electron/js2c/browser_init.js:157:11014)
    at Object.emit (events.js:315:20)__

'Give Feedback' works fine when running tests in Chrome or Edge.

Maybe worth to note I upgraded to Cypress 6.4.0 before using Cypress Studio.

Don't know if Cypress Studio is not intended to be used with Electron. If so perhaps you could add the use of Chrome or Edge to the prerequisites of the Realworld-App

**Reply by panzarino** [reply]
Thanks for the feedback - our intention is definitely to support electron and we'll try to fix this bug asap

Would you be able to [file an issue](https://github.com/cypress-io/cypress/issues/new?assignees=&labels=experiment%3A%20studio&template=1-bug-report.md&title=)?

**Reply by jennifer-shehane** [reply]
This is an issue in 6.4.0 only and will be fixed in our next release. https://github.com/cypress-io/cypress/issues/14912

**Comment by Light07** [comment]
Here are one suspect issue i saw: 
If your test codes contains visual Testing( cypress-plugin-snapshots), then the snapshot compare will fail for sure which blocked the studio to start. 
Cypress Version 6.4.0.
cypress-plugin-snapshots Version: "^1.4.4",

**Reply by panzarino** [reply]
@Light07 Thanks for the report! I'm not sure whether that error would be caused on our end or by the plugin implementation, but I can investigate. There is a decent chance that this might have to be fixed by the plugin author.

**Comment by literalpie** [comment]
I really like this feature. My main feedback is that I would like to see more commands implemented, specifically I think keyboard presses should be recorded so we can do testing of keyboard accessibility. I would like to assume more commands are planned, but I don't see issues in the backlog.

Somewhat related to keyboard testing, it would also be helpful if the testing "frame" was focused when you click the record button. This way, the tab order will start at the right place.

**Reply by panzarino** [reply]
Thanks for the feedback! So keyboard presses is something that could definitely be  implemented fairly easily, my only concern is that it might create a lot of unwanted commands accidentally for users that are not trying to test accessibility/keyboard shortcuts. We could definitely implement something with a toggle switch which would enable the best of both worlds.

Unfortunately, [tabbing is not supported](https://docs.cypress.io/api/commands/type.html#Tabbing) in Cypress so I'm not sure that there would be any benefit to focusing the frame at the start.

**Comment by visortelle** [comment]
Hello! 
I use the studio with react component testing to have ability to run the specific test without making ‘.only()’ changes in code.

This tandem makes Storybook-like solutions almost useless for me. Each test can be considered as an example and it’s definitely better than support “stories” in separated files.  

Hope that https://github.com/evanw/esbuild or similar projects will be more mature and popular and compile time won’t be a problem soon.

P.S. There is a visual bug in cypress@6.5.0. Studio’s magic wand icon becomes invisible in case you have long description string in ‘it()’ statement. Sorry, can’t attach the screenshot now as I write it from phone.

**Reply by panzarino** [reply]
Thanks for the feedback! We're currently making a whole host of changes for component testing that will be going live soon which makes the experience significantly better both on the UI side and compiling/bundling.

**Reply by panzarino** [reply]
Issue for the visual bug - https://github.com/cypress-io/cypress/issues/15182

**Comment by wmedali** [comment]
Hello 
I tried Cypress Studio today, I am using Angular 12 with cypress 6.9.1 installed. 
Trying to save a nominal test but it fails with this error. With no other error in logs. Any idea why this happens ?
<img width="442" alt="Capture d’écran 2021-08-14 à 16 20 59" src="https://user-images.githubusercontent.com/23320355/129449496-0c939bba-864f-4f64-b37c-2a61555505a5.png">

**Reply by panzarino** [reply]
I'll look into it. In the meantime, could you post your `cypress.json` config? It might help me figure out what's going on here

**Reply by wmedali** [reply]
Thanks for the feedback. Here is my` cypress.json` content

```json
{
    "fileServerFolder": ".",
    "fixturesFolder": "./src/fixtures",
    "integrationFolder": "./src/integration",
    "modifyObstructiveCode": false,
    "pluginsFile": "./src/plugins/index",
    "supportFile": "./src/support/index.ts",
    "videosFolder": "../../dist/cypress/apps/extranet-e2e/videos",
    "screenshotsFolder": "../../dist/cypress/apps/extranet-e2e/screenshots",
    "chromeWebSecurity": false,
    "viewportHeight": 800,
    "viewportWidth": 1280,
    "reporter": "../../node_modules/cypress-multi-reporters",
    "reporterOptions": {
        "reporterEnabled": "mochawesome, mocha-junit-reporter",
        "mochaJunitReporterReporterOptions": {
            "mochaFile": "../../dist/cypress/apps/extranet-e2e/reports/junit/results-[hash].xml"
        },
        "mochawesomeReporterOptions": {
            "reportDir": "../../dist/cypress/apps/extranet-e2e/reports",
            "overwrite": false,
            "html": false
        }
    },
    "downloadsFolder": "../../dist/cypress/apps/extranet-e2e/downloads",
    "experimentalStudio" : true
}

**Reply by panzarino** [reply]
I can't seem to figure it out from that - what does your plugins file look like? Are you using a custom preprocessor?

**Reply by wmedali** [reply]
I am using Cypress with angular app. My plugins file : 
const { preprocessTypescript } = require('@nrwl/cypress/plugins/preprocessor');
```
module.exports = (on, config) => {
    // `on` is used to hook into various events Cypress emits
    // `config` is the resolved Cypress config

    // Preprocess Typescript file using Nx helper
    on('file:preprocessor', preprocessTypescript(config));
};
```

**Reply by panzarino** [reply]
I would guess that the preprocessor is not correctly generating sourcemaps. I can look into it a bit but I'm not sure if there's much we can do on our end.

**Comment by khafidmedheb** [comment]
Hi, first of all, thank you for the tool "Cypress Studio" which greatly facilitates the tests. I have one question about the automatisation of the Drag and Drop event in Cypress: Do you know if the future releases of Cypress Studio will be able to deal with the Drag & Drop event justly ?. Thank you.

**Comment by emahuni** [comment]
This is soooo cool.

Add a checkbox, so that when the checkbox is checked it adds not to every assertion that it suggests in the assertions menu.

![image](https://user-images.githubusercontent.com/7290508/137702867-3d50aca5-3e41-4094-91c5-1f4a29c1d818.png)

that way it doesn't have to guess what is being intended.
In the above screenshot, I intend the studio to create assertions that ensure the account # is not invalid.

**Reply by emahuni** [reply]
this can be a basis for other studio assertions or commands build-ups as well.

**Reply by emahuni** [reply]
then add the `not exist` assertion to that list, maybe only when the `not` checkbox is checked.

**Comment by khafidmedheb** [comment]
Hi everyone, Does anyone can manage the Drag & Drop with Cypress ?. I justly installed the plugin "drag-and-drop" but i can't automatise the D&D on my page. Moreover, don't hesitate to reply if you have code examples about automatisation of the D&D with Cypress ;). Thank you :).

**Comment by upwavedave** [comment]
Hello from Texas! I'm having trouble getting Cypress Studio to recognize the **right click** on page elements in the preview area but only on our company's site. On other pages, like the Kitchen Sink test site, it works fine. 

I worry that our app might be hijacking the right-click in some way. Is there a good diagnostic approach I can use to find out why Cypress Studio might not be "seeing" the right-click? Instead of the `Assertions` dropdown, I just get an `Inspect Element` button.

Thanks in advance for any help!

**Comment by khafidmedheb** [comment]
Hi everyone, I want to test a download PDF button with Cypress. During a dowload process, Cy register the file in cypress/downloads. My PDF file to test is registered but i can't parse it with my task command. I got this error message on the test runner "Invalid PDF structure". When i open manually my pdf file in cypess/downloads, i only have this text "This PDF file is encrypted and can't be opened. " on a single page. Does anyone knwo how to fix it ?. Thank you.

**Reply by moosebay** [reply]
I'd recommend preflight.com for this.

**Comment by cosminlupw** [comment]
My application is running inside iframes and I'm using "cypress-iframe": "^1.0.1" to interact with elements inside the page.
Currently, Studio Beta does not record any interactions/simple commands like 'click', 'select', 'etc' with elements in the iframe and has the following output when saving the commands
![image](https://user-images.githubusercontent.com/79325039/151556165-7c27c938-67f3-4c65-926b-ce598d85fa0a.png)
```js
 /* ==== Generated with Cypress Studio ==== */
    
    /* ==== End Cypress Studio ==== */
```
Normally when writing tests I'm using cy.enter("iframe-name", getbody => getBody().find("elementName")) to interact with elements
```js

describe("portal", () => {
  before(() => cy.login());
  it("Verify ", () => {
    cy.enter("iframe-name".then((getBody) => {
      getBody().find("table > tbody").its("length").should("be.greaterThan", 1);
    });
  });
});
```

**Reply by lmiller1990** [reply]
Thanks for this - it would be great to get Studio working with

1. iframes
2. `cy.origin`

I suspect (1) is not too difficult, and (2) is more complex.

**Comment by carwei** [comment]
Would it be possible to always run Cypress Studio in "Add Commands to Test" mode (based on a setting available to the user)? As long as the test ran successfully, I would like to add stuff to it. I'm creating a lot of tests now, and I notice that I have to interrupt the running tests a lot by clicking the magic wand to switch to "edit mode".

**Comment by cb0806151** [comment]
Is there a way to have a `"studioSelects": "xpath"/"class"/"styles"` toggle in the cypress.json? I'm currently getting a test like this when recording in cypress studio and it is unusable since the class names change between page visits. Other than that, great tool and look forward to using it for many projects in the future
```
    it('create account', function() {
        cy.get('._1yxs8iu').check();
        cy.get('._1wy2dtkm').click();
        cy.get('._cg9dizq > ._jro6t0 > ._o8mr7f > img').click();
        cy.get(':nth-child(2) > ._1ymsjqx > ._49qt1p').click();
        cy.get('._1j6otxzf > ._49qt1p').clear();
        cy.get('._1j6otxzf > ._49qt1p').type('BuzzFiz');
        cy.get('._cg9dizq > ._49qt1p').clear();
        cy.get('._cg9dizq > ._49qt1p').type('BuzzFiz2.');
        cy.get('._1ymsjqx > ._49qt1p').click();
        cy.get('._1j6otxzf > ._49qt1p').clear();
        cy.get('._1j6otxzf > ._49qt1p').type('BuzzFiz2');
        cy.get('._cg9dizq > ._49qt1p').clear();
        cy.get('._cg9dizq > ._49qt1p').type('BuzzFiz2.');
        cy.get('._1wy2dtkm').click();
        cy.get(':nth-child(2) > ._dl3obf > ._1eiutku > :nth-child(2)').click();
        cy.get('._d9bqdc > ._t2mr9i').click();
        cy.get('._27goqx > :nth-child(2)').click();
        cy.get(':nth-child(3) > ._dl3obf > ._1eiutku > ._179xpb9').click();
        cy.get('#\\30 ').clear();
        cy.get('#\\30 ').type('lumber');
        cy.get('._u02imw > div').click();
        cy.get('._49qt1p').clear();
        cy.get('._49qt1p').type('BuzzFiz2.');
        cy.get(':nth-child(1) > ._1wy2dtkm > ._u02imw > div').click();
        /* ==== End Cypress Studio ==== */
    });
   ```

**Reply by pstakoun** [reply]
Have you tried adding `data-*` selectors for these elements? Here's a [blog post](https://www.cypress.io/blog/2022/08/30/how-to-use-studio-in-cypress-10-7#a-note-on-resilient-test-selectors) with some helpful info.

**Comment by GregFinzer** [comment]
I have a fictitious Order page that has all the HTML5 controls on it:
http://samssoups.kellermanstores.com/Order.html

Here are controls that are not automated and filled correctly:
input type="range"
input type="color"

**Reply by lmiller1990** [reply]
👀 

Thanks for this, good point - we need more support for all the different `input type="..."`. I'd also like to see support for `input type="date"`.

**Comment by SheaBelsky** [comment]
How can I configure Cypress Studiot to look for a file within a specific subdirectory?

My test file runs from `cypress/e2e/jobs/job_profile/editable.ts`. When I add interactions with Cypress Studio, I get the following error:

```
ENOENT: no such file or directory, open '/Users/sheabelsky/Documents/mentra/mentra-platform/e2e/jobs/job_profile/editable.ts'

Cypress was unable to save these commands to your spec file. You can use the copy button below to copy the commands to your clipboard.
```

It looks like Cypress naively assumes the est file will always be executed from the root of the project, but my tests are executed from within a `cypress` subdirectory. I don't see a clear way to configure Cypress Studio to start from this folder - what do I need to do to configure this?

**Reply by warrensplayer** [reply]
Hi @SheaBelsky, sorry for the delayed response.  Can you share your cypress config file?

**Reply by SheaBelsky** [reply]
@warrensplayer This is my `cypress.config.ts` file:

```js
import { defineConfig } from "cypress";

export default defineConfig({
  defaultCommandTimeout: 10000,
  env: {
    // Env removed
  },
  projectId: "removed",
  e2e: {
    // We've imported your old cypress plugins here.
    // You may want to clean this up later by importing these.
    setupNodeEvents(on, config) {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      return require("./plugins/index.ts")(on, config);
    },
    baseUrl: "http://localhost:3100",
    specPattern: "cypress/e2e/**/*.{js,jsx,ts,tsx}"
  }
});

```

**Reply by warrensplayer** [reply]
@SheaBelsky Thanks for providing this.  I have not been able to reproduce your error.  Can you please file an issue by selecting the "Get Started" button for a "Bug Report" here: https://github.com/cypress-io/cypress/issues/new/choose Please fill out all the sections as best as possible and include your config file above.  Also, if you are able to create a simple project that reproduces the problem, please share that as well.  Thanks again!

**Comment by ovirta** [comment]
Thank you for this helpful tool!

Just wondering if there is a way to auto-generate intercepted routes via Cypress Studio? 

Meaning that when adding commands to the spec I could right click a link and an intercepted route is generated by the studio.

ChatGPT is suggesting ("Right click on an element that you want to assert on or intercept a request from.") that there is such a feature but all that I can find is the normal right-click Assertion options. Or maybe I misunderstood something.

**Comment by lmiller1990** [comment]
> Thank you for this helpful tool!
> ChatGPT is suggesting ("Right click on an element that you want to assert on or intercept a request from.") that there is such a feature but all that I can find is the normal right-click Assertion options. Or maybe I misunderstood something.

No such feature exists in Studio at this point. Can you explain more about how you would like this to work? Eg: I'm understanding you want to click a button and it launches an API request, and you'd like to intercept this. Do you want to return some mock data? Eg, you'd like to see a UI that lets you chose a fixture, or enter some test data, as part of the Studio flow?

**Reply by ovirta** [reply]
When creating integration tests for our frontend I find it tedious and time consuming to create intercepted requests/responses. For instance clicking a button or navigating to a new page can in worst case create dozen backend calls to different service apis. After making changes in the ui you make new requests to same apis with modified request and response data.

Ideally when clicking a button or navigating I could (right click) choose to create a named intercepted requests with fixtures for all api calls that take place after the event.

**Reply by lmiller1990** [reply]
Thanks for the reply - can I probe you for some more information?

> tedious and time consuming to create intercepted requests/responses

Agreed, sounds tedious.

> (right click) choose to create a named intercepted requests with fixtures for all api

It sounds like you'll still need to go through the tedious process of making those intercepts, though -- If I'm understanding correctly, you aren't really looking to change the process and end result (the generated test) but just have a UI that makes it faster / better?

If you have any ideas (even a drawing made in Paint, or something) feel free to share it here. I'm keeping an eye on all the feedback / feature requests here in the meantime, and figuring out how they would be implemented / how much work / value they will have relative to each other.

**Reply by atulpatildbz** [reply]
Would love to see this feature in cypress studio.

I work on an application that has a lot of slow APIs. In some cases we need to wait for APIs to complete and succeed before proceeding.
Adding this manually is slightly tedious.

**Comment by abinavanil-11** [comment]
Hey team,

As an automation developer with two years of experience in Cypress, I recently explored the idea of involving non-technical testers in test automation using Cypress Studio. While I appreciate the tool's potential, I've identified a few drawbacks that I believe need consideration and improvement:

### Inconsistent Element Selection:
Cypress Studio uses a structure from the outermost HTML element to the innermost for element selection. This approach can result in inconsistency and might not be the most robust method for all scenarios.

### Dynamic Selectors and Coordinates:
Cypress Studio occasionally utilizes dynamic selectors that change with each run and coordinates as selectors, creating inconsistency, especially when dependent on window size. A more stable approach to element selection could enhance the reliability of tests.

### Default Wait Time and API Waits:
Cypress Studio lacks a default wait time, and automatic API waits are not included in the generated code. Running code without manually adding waits may lead to failures, particularly when there's no wait for the required page to load. Including better support for handling wait times in the generated code would significantly improve the tool's usability.

I believe addressing these concerns would make Cypress Studio more user-friendly for non-technical testers. I appreciate your consideration of these suggestions and look forward to seeing how Cypress Studio evolves. These proposed changes aim to improve the user experience, reduce potential sources of test failure, and ultimately make Cypress Studio a more effective and user-friendly tool for non-technical testers. Your consideration of these enhancements is crucial in achieving a more streamlined and reliable automated testing process.

**Comment by ovirta** [comment]
> Thanks for the reply - can I probe you for some more information?

Of course and now that I am back to writing cypress tests again and generating non-trivial amount of intercepted api calls it is time to provide it.

Playing around with browser dev-tools realized that it is possible to generate intercepts from the output of "Copy all as HAR". Would love to have a feature in Cypress that would allow me to generate code for the intercepted calls from this "HTTP Archive Format" log file.

Work flow would involve:
- refresh test page
- generate har log file
- create stub cy.ts file using har information with intercept code snippets for each request in the stub file (maybe even with json files imported from ../fixtures/)

Hope this clarifies the request (pun intended) bit more :)

**Comment by robbaartwijk** [comment]
Hi,

Very nice new feature, I will test it the coming days.

One suggestion : It would be fantastic if one could select an entry from an autocomplete. So .type('somevalue'), the list appears, and then select an option from the list, which is then added to the test script.

Kind regards,

Rob.

**Comment by fangqi5** [comment]
Hi Team This is a great feature. 
Could you please consider exposing Studio events to developers in setupNodeEvents or Cypress.on? 
This will help developers expand their custom recording UI and the final generated file content.

**Reply by pstakoun** [reply]
Hey @fangqi5, quick product question: what are some of the specific things you are looking to do in terms of customization?

**Reply by fangqi5** [reply]
> Hey @fangqi5, quick product question: what are some of the specific things you are looking to do in terms of customization?

I'd like to enhance the assertion panel by implementing support for cypress-if syntax and adding custom functionality, which would streamline the assertion workflow for developers

**Comment by jennifer-shehane** [comment]
Hi, I wanted to share a blurb about an upcoming webinar we're having that will go into more detail about our future plans for Cypress Studio, for those interested.

>Join us on January 28, 2025, 12-1pm EST for an exclusive preview of Cypress AI and the exciting updates coming to Cypress Studio! 
>
>Discover how AI-powered Test Generation creates targeted tests in a single click, eliminating manual effort while aligning with your team’s unique testing patterns. Plus, get a sneak peek at new Studio enhancements that make test authoring faster, smarter, and more intuitive. 
>
>Don’t miss this chance to experience the future of intelligent test creation—register now to save your spot or receive the recording: [Register here](https://cypress.registration.goldcast.io/webinar/4f300089-6435-4991-885b-2ec4285d0cd7?&utm_medium=studio-discussion-14339&utm_content=register-here&utm_source=github).

**Reply by agu-test** [reply]
Hello Jennifer, 
Is it possible to receive the recording? I signed up today and can enter the event, but I cannot find the recording.
Many thanks in advance!
Kind regards

**Comment by jennifer-shehane** [comment]
Hi everyone! We've made some big updates in Cypress 15 for Studio. You can see a demo video below and read all about it in the [Cypress 15 blog post](https://www.cypress.io/blog/cypress-15-the-foundation-for-whats-next). 

This is more than a quality-of-life update. It’s the beginning of our vision for AI-assisted test creation. The new Studio experience gives us the building blocks to experiment, iterate, and eventually ship smarter test tooling that learns from your app and your tests. [Sign up for early access to Studio AI.](https://on.cypress.io/studio-ai-waitlist?utm_source=github&utm_medium=studio-experimental-feedback-discussion&utm_content=sign-up-for-early-access-to-studio-ai)

To try the new Studio, turn on `experimentalStudio` in the `e2e` config of your Cypress configuration file. Play with it. Break it. Tell us what you love and what you don’t.

https://github.com/user-attachments/assets/eeaf616e-1ed3-4cd1-856b-598feb4e11e9

**Reply by vanessaolmos88** [reply]
hi modify the version of node to 20, upped version of linux, do we have compatibility or do we need to make adjustments in test cases? now I get so running.
 [3946:0826/134650.825671:ERROR:dbus/bus.cc:408] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[3946:0826/134702.039129:ERROR:dbus/bus.cc:408] Failed to connect to the bus: Failed to connect to socket ¿)
[3946:0826/134702.203920:ERROR:dbus/object_proxy.cc:590] Failed to call method: org.freedesktop.DBus.NameHasOwner: object_path= /org/freedesktop/DBus: unknown error type:  
 And when you raise the graphic part is not activated
<img width="917" height="617" alt="Captura de pantalla 2025-08-26 093316" src="https://github.com/user-attachments/assets/50c733ba-8946-4ca5-92d7-019f9213fdbc" />

**Reply by jennifer-shehane** [reply]
@vanessaolmos88 Hi, tests run with cucumber are not able to be displayed in Studio - this is the error you are seeing. 

Your first issue about this output in the terminal is [this issue](https://github.com/cypress-io/cypress/issues/32290) - which we're working on.

**Comment by ismailcendik** [comment]
The record system is not working stably. After recording one or two steps, it stops capturing. When I click the record button again and restart it, the recording continues from where it left off.

https://github.com/user-attachments/assets/1dcfc879-41d9-47e2-8636-727c20a3a3a2

**Comment by goska-malaga** [comment]
well, I actually had to change flag to false since there is  RequestError: Error: unable to get local issuer certificate
it is somewhat weird since the tests can be run so I have no idea what should I do about it
after restarting there is still button +New Test on hover however it doesn't work - I assume this button should not show up at all

**Comment by Kees0** [comment]
I really miss this function in v15 (left icon):
<img width="101" height="79" alt="image" src="https://github.com/user-attachments/assets/c10a62c2-4612-4f68-af43-efb024c7bdc6" />

**Comment by haykevin** [comment]
I have been playing with ExperimentalStudio in v15 and I like a lot about it, but I am suffering because of one feature that is no longer there. I don't know all the names of the tools, but in the Cypress Test Runner pre-15, up at the top, to the left of the URL there was that compass-looking icon that I could click and try and hone my selectors by filling out cy.get. That gave immediate feedback as I typed. I used that all the time to show elements that were being matched. If my script errored out, I could copy my bad selector, paste it into that cy.get area and play with it until it highlighted my element. I don't know how to replace the functionality of that compass icon in v15 Experimental Studio.

The few videos I can find about Studio Beta only show clicking elements on the web page, and right-clicking them to make assertions.

For now I have to downgrade to 14 but I am looking forward to seeing how this is supposed to be done now.

**Comment by jennifer-shehane** [comment]
For followup on the selector playground functionality - add your feedback here: https://github.com/cypress-io/cypress/issues/32672

This discussion will be closed due to Studio no longer being experimental.
